### DISCUSSION TITLE
How to edit the road surfaces in the map with UE4 or other methods

### DISCUSSION THREAD

**Discussion by Lcp9202** [carla-simulator/carla#8009] [Q&A]
Currently, I am using CARLA to collect radar simulation data for urban roads. However, the roads are very smooth, and I wish to edit the road surfaces using UE4 to create a more realistic, uneven terrain. This requires the source files, as without them, I am unable to make any modifications. So is it possible to provide source files for UE4 development? Or if there any other methods to edit the road surfaces?

**Comment by o-antill** [comment]
You can get the source code for Carla following this guide "https://carla.readthedocs.io/en/latest/build_linux/". It will also show you how to install Carla's version of UE4 4.26. Note that when you apply to get the version of Unreal engine from the EpicGames repo you will need to of first linked your Github account with your EpicGames account (Make sure you do this using the same email for both).

**Reply by o-antill** [reply]
Once you have the source code look for resources like RoadRunner to make custom maps!

**Reply by Lcp9202** [reply]
Thank you for your reply. My system is windows and I have installed carla and Unreal Engine 4.26 as instructed before asking the question, but I can't seem to successfully open the uproject project, prompting me (1) "The following modules are missing or bulit with a different engine version:  CarlaUE4
Carla
CarlaTools
HoudiniEngine
HoudiniEngineEditor
HoudiniEngineRuntime
StreetMapRuntime
StreetMaplmporting
Would you like to rebuild them now?" At this point, I'm sure I'm using UE4.26, and when I select "rebuild", I get a prompt (2) "CarlaUE4 could not be compiled. Try rebuilding from source manually" 。 So I wondered if I needed a so-called "source" file

**Reply by o-antill** [reply]
unfortunately i have never had the windows version working however i know people have been able to. However to save yourself a lot of time i look into maybe trying to dual boot a linux system along side and working with that! Not saying it is not possible by any means but in terms of saving time and complexity

**Reply by Lcp9202** [reply]
Thanks for your reply. I would consider using Linux. By the way, have you tried to edit a map successfully on Linux before?

**Reply by o-antill** [reply]
Sorry for the late reply i have edited the map in the sense of adding feature like Road side tags but never the actual architecture of the street or surrounding buildings!

**Reply by Lcp9202** [reply]
How do you open the file and edit it, I can't open it right now

**Reply by o-antill** [reply]
do you have the unreal editor installed ? The carla specific version ?

**Comment by Lcp9202** [comment]
yes,I download the UE，the version is 4.26

**Reply by o-antill** [reply]
so for example for my case i added a plane and then added an image to it 
![Screenshot from 2024-08-07 11-20-33](https://github.com/user-attachments/assets/5e50a03b-5b83-4c91-8e6e-dd043dfff79e)

**Reply by o-antill** [reply]
this is a tag i am reading from the camera data supplied by and RGB camera i attached to my vehicle

**Reply by Lcp9202** [reply]
I konw what you mean,but now I can’t open the  file. I ask another question at here: https://github.com/carla-simulator/carla/discussions/8025#discussion-7008618

**Reply by Lcp9202** [reply]
> so for example for my case i added a plane and then added an image to it ![Screenshot from 2024-08-07 11-20-33](https://private-user-images.githubusercontent.com/171234048/355775620-5e50a03b-5b83-4c91-8e6e-dd043dfff79e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMyNzQ0NTUsIm5iZiI6MTcyMzI3NDE1NSwicGF0aCI6Ii8xNzEyMzQwNDgvMzU1Nzc1NjIwLTVlNTBhMDNiLTViODMtNGM5MS04ZTZlLWRkMDQzZGZmZjc5ZS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODEwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxMFQwNzE1NTVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MjI4YWIyYWRlODhlN2VlYjRkOGNjMTgzM2FiYzU3OTZlOWFlZWIyMjhmMjY3OTk1NjMzNDQ4ZGQ1NmY3ZjdhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.b1-ZLeiHvVURLEoDtSF8BJjEu7O7N2_EHZ2Lyv2LgXU)

And when you added the picture upon the road,if you use the camera or lidar to scan it,could you get the information just like the distance,width and height of the picture or the point cloud?

**Reply by o-antill** [reply]
i didnt test it with a lidar but with a camera and using OpenCV i got it working
